src/highlight/file_tester.rs

```diff
@@ -391,27 +393,83 @@ pub fn is_potential_cd_path(
 ///     false: the filesystem is not case insensitive
 ///     true: the file system is case insensitive
 pub type CaseSensitivityCache = HashMap<WString, bool>;
+
 fn fs_is_case_insensitive(
     path: &wstr,
     fd: RawFd,
     case_sensitivity_cache: &mut CaseSensitivityCache,
 ) -> bool {
-    let mut result = false;
-    if *_PC_CASE_SENSITIVE != 0 {
-        // Try the cache first.
+    if let Some(name) = *_PC_CASE_SENSITIVE {
         match case_sensitivity_cache.entry(path.to_owned()) {
-            Entry::Occupied(e) => {
-                /* Use the cached value */
-                result = *e.get();
+            Entry::Occupied(e) => *e.get(),
+            Entry::Vacant(e) => {
+                // fpathconf: 0 -> insensitive, 1 -> sensitive, -1 -> ошибка/неизвестно
+                let ret = unsafe { libc::fpathconf(fd, name) };
+                let v = ret == 0;
+                e.insert(v);
+                v
             }
+        }
+    } else {
+        match case_sensitivity_cache.entry(path.to_owned()) {
+            // Ask the system. A -1 value means error (so assume case sensitive), a 1 value means case
+            // sensitive, and a 0 value means case insensitive.
+            Entry::Occupied(e) => *e.get(),
             Entry::Vacant(e) => {
-                // Ask the system. A -1 value means error (so assume case sensitive), a 1 value means case
-                // sensitive, and a 0 value means case insensitive.
-                let ret = unsafe { libc::fpathconf(fd, *_PC_CASE_SENSITIVE) };
-                result = ret == 0;
-                e.insert(result);
+                let v = pure_rust_case_insensitive_at_fd(fd);
+                e.insert(v);
+                v
             }
         }
     }
-    result
+}
+
+/// Fallback: create a file called "lower", check if it's visible as "UPPER".
+/// Returns true if the FS is case-insensitive.
+#[cfg(any(unix, target_os = "wasi"))]
```

Comment:
is this relevant for wasi platforms?
Do these even have file system access?
Can fish run on wasi?